8258010: Debug build failure with clang-10 due to -Wdeprecated-copy

shqking opened the pull request:
1. '-Wdeprecated-copy'
As specified in C++11 [1], "the generation of the implicitly-defined
copy constructor is deprecated if T has a user-defined destructor or
user-defined copy assignment operator". The rationale behind is the
well-known Rule of Three [2].

Introduced since gcc-9 [3] and clang-10 [4], flag '-Wdeprecated-copy'
warns about the C++11 deprecation of implicitly declared copy
constructor and assignment operator if one of them is user-provided.
Defining an explicit copy constructor would suppress this warning.

The main reason why debug build with gcc-9 or higher succeeds lies in
the inconsistent warning behaviors between gcc and clang. See the
reduced code example [5]. We suspect it might be return value
optimization/copy elision [6] that drives gcc not to declare implicit
copy constructor for this case.

Note that flag '-Wdeprecated' in clang-8 and clang-9 would also raise
warnings for deprecated defintions of copy constructors. However,
'-Wdeprecated' is not enabled by '-Wall' or '-Wextra'. Hence, clang-8
and clang-9 are not affected.

~~2. '-Wimplicit-int-float-conversion'~~
~~Making the conversion explicit would fix it.~~

~~Flag '-Wimplicit-int-float-conversion' is first introduced in clang-10.~~
~~Therefore clang-8 and clang-9 are not affected. The flag with similar~~
~~functionality in gcc is '-Wfloat-conversion', but it is not enabled by~~
~~'-Wall' or '-Wextra'. That's why this warning does not apprear when~~
~~building with gcc.~~

[1] https://en.cppreference.com/w/cpp/language/copy_constructor
[2] https://en.cppreference.com/w/cpp/language/rule_of_three
[3] https://www.gnu.org/software/gcc/gcc-9/changes.html
[4] https://releases.llvm.org/10.0.0/tools/clang/docs/ReleaseNotes.html
[5] https://godbolt.org/z/err4jM
[6] https://en.wikipedia.org/wiki/Copy_elision#Return_value_optimization


Note that we have tested with this patch, debug build succeeded with clang-10 on Linux X86-64/AArch64 machines.
Note that '--with-extra-cxxflags=-Wno-implicit-int-float-conversion' should be added when configuration. It's another issue (See JDK-8259288)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258010](https://bugs.openjdk.java.net/browse/JDK-8258010): Debug build failure with clang-10 due to -Wdeprecated-copy


### Reviewers
 * [Xin Liu](https://openjdk.java.net/census#xliu) (@navyxliu - no project role) ⚠️ Review applies to 68679966d8df8729602d5fc9ca6421ec5abdc75c
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1874/head:pull/1874`
`$ git checkout pull/1874`

shqking commented:
/label add build

kimbarrett commented:
I think the two issues described here are distinct and should be dealt
with in separate bugs and PRs.  Their only relation is that both arise
with using clang-10.  But they are very different problems, in very
different parts of the code, and probably ought to be reviewed by
folks from different teams.

kimbarrett commented on src/hotspot/share/opto/node.hpp:
```diff
@@ -1452,6 +1455,8 @@ class DUIterator_Last : private DUIterator_Fast {
   DUIterator_Last() { }
   // initialize to garbage
 
+  DUIterator_Last (const DUIterator_Last& that) : DUIterator_Fast(that) {}
```
Since DUIterator_Last is just delegating both the copy constructor and
assignment operator to the base class, their copy constructor and
assignment operator would be better as the default (either implicit or
explicit using `=default`) rather than explicit code.

navyxliu replied:
Agree.  c2 actually never uses the assignment operator of DUIterator_Last.  It needs the copy constructor in this line.
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.hpp#L1499

you can delete the following code or leave it =default.
```
-  void operator=(const DUIterator_Last& that)
-    { DUIterator_Fast::operator=(that); }
```

shqking replied:
Thanks for your comments.
Will remove the assignment operator of DUIterator_Last.

shqking commented:
> I think the two issues described here are distinct and should be dealt
> with in separate bugs and PRs. Their only relation is that both arise
> with using clang-10. But they are very different problems, in very
> different parts of the code, and probably ought to be reviewed by
> folks from different teams.

Thanks for your comment.

Warning message of '-Wimplicit-int-float-conversion' was further encountered after we fixed the build failure caused by '-Wdeprecated-copy' first. That's why we put them in one PR initially.

Yes. Your way is much better. But we suppose the issue of '-Wimplicit-int-float-conversion' is trivial and putting them in separate PRs might raise another internal review process (for our side) by which extra time is needed. I was wondering could we continue in one single PR. :)

navyxliu reviewed (approved):
LGTM. It still needs other's approval.

shqking commented:
> _Mailing list message from [Kim Barrett](mailto:kim.barrett@oracle.com) on [build-dev](mailto:build-dev@openjdk.java.net):_
> 
> > On Dec 22, 2020, at 8:52 PM, Hao Sun <github.com+16932759+shqking at openjdk.java.net> wrote:
> > 1. '-Wdeprecated-copy'
> > As specified in C++11 [1], "the generation of the implicitly-defined
> > copy constructor is deprecated if T has a user-defined destructor or
> > user-defined copy assignment operator". The rationale behind is the
> > well-known Rule of Three [2].
> > Introduced since gcc-9 [3] and clang-10 [4], flag '-Wdeprecated-copy'
> > warns about the C++11 deprecation of implicitly declared copy
> > constructor and assignment operator if one of them is user-provided.
> > Defining an explicit copy constructor would suppress this warning.
> > The main reason why debug build with gcc-9 or higher succeeds lies in
> > the inconsistent warning behaviors between gcc and clang. See the
> > reduced code example [5]. We suspect it might be return value
> > optimization/copy elision [6] that drives gcc not to declare implicit
> > copy constructor for this case.
> 
> C++17 "guaranteed copy elision" is implemented in gcc7.
> http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0135r1.html
> 

Thanks for your comment.
Initially we only suspected it might be copy elision that made gcc and clang to behave differently on this warning, and we were not aware of this flag '-fno-elide-constructors'. 
Thank you for pointing it out.

> Using -fno-elide-constructors makes it obvious that the deprecated
> implicit copy constructors are indeed being elided, so no deprecation
> warning.
> 

I suppose you want to express 'Using -**felide-constructors**' here.
gcc with '-fno-elide-constructos' would produce derepcated warnings for this issue as clang-10 does.

> Why doesn't this patch similarly define DUIterator copy constructor?
> It seems to have the same issue, and I'm surprised clang-10 didn't
> complain about it too. Certainly gcc with -fno-elide-constructors
> complains about the defaulted implicit constructor.
> 

I'm afraid we have noticed the same issue for 'DUIterator' before.
Yes. It's weird that clang-10 didn't raise warning here. ( verified it on my local machine.)

> But I discovered something alarming while experimenting. Building
> with gcc10.2 with -fno-elide-constructors doesn't seem to be possible.
> I get different kinds of failures depending on how DUIterator is
> defined:
> 
> - implict: deprecation warning (as expected)
> - =delete: error, deleted function used
> - =default: assert in os::free
> - _idx and reset from that: assert in reset
> 
> Without -fno-elide-constructors, all of the variants seem to work
> except =delete, which still fails because the deleted function is
> used. (I didn't test the "working" cases extensively though.)
> 
> So there's something problematic, though I don't understand the code
> well enough to understand what.

Thanks for your tests. 
But I have no idea how to fix it right now either.
Do you know anyone who is familiar with these code and maybe we can invite him/her to help take a look at this issue?
Thanks.

> 
> Interestingly, some of the complexity and weirdness around copy/assign
> for these classes may be an attempt at providing something like move
> semantics in a C++98 code base. I've noticed a number of places in
> HotSpot that are doing that.
> 
> Defining DUIterator_Fast and DUIterator_Last as movable but not
> copyable (explicitly delete the copy constructor and copy assign
> operator, and define the move constructor and move assign operator
> with the reset) works, even with -fno-elide-constructors.

shqking commented:
> _Mailing list message from [Kim Barrett](mailto:kim.barrett@oracle.com) on [build-dev](mailto:build-dev@openjdk.java.net):_
> 
> > On Dec 24, 2020, at 3:44 PM, Xin Liu <xliu at openjdk.java.net> wrote:
> > On Thu, 24 Dec 2020 17:27:39 GMT, Kim Barrett <kbarrett at openjdk.org> wrote:
> > > [?]
> > > Since DUIterator_Last is just delegating both the copy constructor and
> > > assignment operator to the base class, their copy constructor and
> > > assignment operator would be better as the default (either implicit or
> > > explicit using `=default`) rather than explicit code.
> > 
> > 
> > Agree.  c2 actually never uses the assignment operator of DUIterator_Last.  It needs the copy constructor in this line.
> > https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.hpp#L1499
> > you can delete the following code or leave it =default.
> > -  void operator=(const DUIterator_Last& that)
> > -    { DUIterator_Fast::operator=(that); }
> 
> DUIterator_Last::operator= *is* used, for example:
> https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.hpp#L1473
> 
> That doesn't change whether defaulting DUIIterator_Last's copy constructor
> and copy assign works though (whether implicit or explicit). So making it
> implicit does work.
> 
> I think making it explicitly defaulted might be better though, to make it
> clear the default behavior is intentional and it wasn't accidentally left as
> the implicit default. This is because the default isn't right for the other
> related classes.

Yes. You're right. It's much better to make it explicitly defaulted.
May I have your opinion @navyxliu ?

shqking commented:
> _Mailing list message from [Kim Barrett](mailto:kim.barrett@oracle.com) on [build-dev](mailto:build-dev@openjdk.java.net):_
> 
> > On Dec 29, 2020, at 10:33 PM, Hao Sun <github.com+16932759+shqking at openjdk.java.net> wrote:
> > > _Mailing list message from [Kim Barrett](mailto:kim.barrett at oracle.com) on [build-dev](mailto:build-dev at openjdk.java.net):_
> > > C++17 "guaranteed copy elision" is implemented in gcc7.
> > > http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0135r1.html
> > 
> > 
> > Thanks for your comment.
> > Initially we only suspected it might be copy elision that made gcc and clang to behave differently on this warning, and we were not aware of this flag '-fno-elide-constructors'.
> > Thank you for pointing it out.
> > > Using -fno-elide-constructors makes it obvious that the deprecated
> > > implicit copy constructors are indeed being elided, so no deprecation
> > > warning.
> > 
> > 
> > I suppose you want to express 'Using -**felide-constructors**' here.
> > gcc with '-fno-elide-constructos' would produce derepcated warnings for this issue as clang-10 does.
> 
> I really did mean "-fno-elide-constructors". The idea is that having the
> normally working build fail with "-fno-elide-constructors" provides evidence
> for the "working because of copy elision" conjecture.
> 
> clang has supported -f[no-]elide-constructors for a while.
> 
> It appears that either clang is different from gcc for -felide-constructors
> (which seems unlikely), or clang (clang-10) is doing the deprecation warning
> at a different point from gcc (quite plausible). That is, clang could be
> checking for the deprecated case before eliding the call, while gcc is
> checking for the deprecated case after copy elision has been applied.

Thanks for your reply.
I checked the source code of clang-10 and gcc-9 and found that:

1) for clang-10,
'Wdeprecated-copy' is implemented at the 'Sema' module of clang. See https://github.com/llvm/llvm-project/blob/release/10.x/clang/lib/Sema/SemaDeclCXX.cpp#L13585

Flag 'felide-constructors' would enable 'felide_constructors' and flag 'fno-elide-constructors' would enables 'fno_elide_constructors'. (See https://github.com/llvm/llvm-project/blob/release/10.x/clang/include/clang/Driver/Options.td).
Then 'ElideConstructors' will be set (See https://github.com/llvm/llvm-project/blob/release/10.x/clang/lib/Frontend/CompilerInvocation.cpp#L2863)
Finally, constructors might be elided in several spots in 'CodeGen' module. 
See: 
https://github.com/llvm/llvm-project/blob/release/10.x/clang/lib/CodeGen/CGStmt.cpp#L1094
https://github.com/llvm/llvm-project/blob/release/10.x/clang/lib/CodeGen/CGExprCXX.cpp#L611
https://github.com/llvm/llvm-project/blob/release/10.x/clang/lib/CodeGen/CGDecl.cpp#L1405

As far as I know, 'Sema' module is conducted to check semantics errors before 'CodeGen' module.
That verified your conjecture, i.e. 'clang could be checking for the derepcated case before eliding the call'.

2) for gcc-9,
'felide-constructors' and 'Wdeprecated-copy' are implemented in a number of spots in gcc. I currently didn't figure out their execution order clearly.

But in one of the use points at function build_over_call(), 'flag_elide_constructors' (See https://github.com/gcc-mirror/gcc/blob/releases/gcc-9/gcc/cp/call.c#L8538) is handled **before** 'warn_deprecated_copy' (See https://github.com/gcc-mirror/gcc/blob/releases/gcc-9/gcc/cp/call.c#L8608 and https://github.com/gcc-mirror/gcc/blob/releases/gcc-9/gcc/cp/call.c#L8679)

Hope that my finding is helpful. Thanks.

kimbarrett commented on src/hotspot/share/opto/node.hpp:
```diff
@@ -1392,6 +1392,9 @@ class DUIterator_Fast : public DUIterator_Common {
     { /*initialize to garbage*/         debug_only(_vdui = false); }
 
+  DUIterator_Fast(const DUIterator_Fast& that)
+    { _outp = that._outp;               debug_only(reset(that)); }
```
`reset` tests `_vdui`, but nothing here has set it, so it's uninitialized and that test wil be UB.

I'm also not sure why it's okay for `operator=` to use whatever the current value of `_vdui` might be; that could leave `_last` as the old value rather than refreshing from `that`, which seems wrong.  This is aabout pre-existing code that looks questionable to me.

shqking replied:
I suppose the constructor would be invoked before the copy assignment operator. That is `_vdui` gets initialized already in the ctor `DUIterator_Fast()` for `operator=` case. Right?

Yes. For our newly-added copy constructor for `DUIterator_Fast`, we should initialize `_vdui` as `false`. It may be defined as below.
```
DUIterator_Fast(const DUIterator_Fast& that)
    { _outp = that._outp;               debug_only(_vdui = false; reset(that)); }
```

kimbarrett replied:
That's true on the first assignment of `t2`. But what if `t2` is reassigned
to some other iterator. That assignment sees `_vdui` true, and keeps the old
value of `_last` rather than updating the value from that other iterator. Is
that really correct? It certainly seems strange. I'm trying to find someone
who understands this code to get involved, but holidays.

shqking replied:
Thanks for your explanation. Yes, you're right. I didn't realize the re-assignment scenario. 

However, I noticed that there does NOT exist such re-assignment cases. 
Simply grep `= DUIterator_Last(` and `= DUIterator_Fast(`.
Is there anything I missed?

shqking replied:
@vnkozlov I was wondering if you could take a look at this? We're not sure whether 'operator=' is problematic or not. Thanks.

shqking replied:
I manually checked the usages of assignment operators for class DUIterator, DUIterator_Fast and DUIterator_Last. (Simply grep the class names in the source code and check the context). 

I found there exist only a couple of re-assignment usages. However, I guess kind of reset operations are conducted in these sites, and I suspect  the implementation of `operator=` might be good.

As I examined, only the following 3 sites are found where re-assignment happens
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/cfgnode.cpp#L565
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/phaseX.cpp#L1878
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/split_if.cpp#L452

Please take `cfgnode.cpp` as an example. `j` is first assigned at line 568 and it would be assigned again if flag `progress` is true. However, I suppose `j` gets reset at line 578 in such case.

It might be incorrect if I missed something.

vnkozlov replied:
Code is correct. _last should not be updated with additional assignments which updates only node's information.

shqking replied:
Thanks a lot for your explanation @vnkozlov .

navyxliu commented:
> > _Mailing list message from [Kim Barrett](mailto:kim.barrett@oracle.com) on [build-dev](mailto:build-dev@openjdk.java.net):_
> > > On Dec 24, 2020, at 3:44 PM, Xin Liu  wrote:
> > > On Thu, 24 Dec 2020 17:27:39 GMT, Kim Barrett  wrote:
> > > > [?]
> > > > Since DUIterator_Last is just delegating both the copy constructor and
> > > > assignment operator to the base class, their copy constructor and
> > > > assignment operator would be better as the default (either implicit or
> > > > explicit using `=default`) rather than explicit code.
> > > 
> > > 
> > > Agree.  c2 actually never uses the assignment operator of DUIterator_Last.  It needs the copy constructor in this line.
> > > https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.hpp#L1499
> > > you can delete the following code or leave it =default.
> > > 
> > > * void operator=(const DUIterator_Last& that)
> > > * { DUIterator_Fast::operator=(that); }
> > 
> > 
> > DUIterator_Last::operator= _is_ used, for example:
> > https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.hpp#L1473
> > That doesn't change whether defaulting DUIIterator_Last's copy constructor
> > and copy assign works though (whether implicit or explicit). So making it
> > implicit does work.
> > I think making it explicitly defaulted might be better though, to make it
> > clear the default behavior is intentional and it wasn't accidentally left as
> > the implicit default. This is because the default isn't right for the other
> > related classes.
> 
> Yes. You're right. It's much better to make it explicitly defaulted.
> May I have your opinion @navyxliu ?
Sorry, I overlooked that copy assignment case.
Making it explicitly default is better to me too. no objection!

shqking commented:
Thanks for your comments. @kimbarrett and @navyxliu 
I updated the patch based on my understanding. Please check the latest commit.

As @kimbarrett mentioned, I suppose there still exist the following problems to be address.
1) why clang-10 with '-Wdeprecated-copy' does NOT raise warning for class DUIterator. It's weird.
2) the assert failure when building with gcc '-fno-elide-constructors'. Might not be related to our patch.
3) the implementation of 'operator=' for class DUIterator_Fast might be problematic.

vnkozlov reviewed (approved):
node.hpp changes seems fine.
Passed tier1 builds and testing.

shqking commented:
> > I think the two issues described here are distinct and should be dealt
> > with in separate bugs and PRs. Their only relation is that both arise
> > with using clang-10. But they are very different problems, in very
> > different parts of the code, and probably ought to be reviewed by
> > folks from different teams.
> 
> Thanks for your comment.
> 
> Warning message of '-Wimplicit-int-float-conversion' was further encountered after we fixed the build failure caused by '-Wdeprecated-copy' first. That's why we put them in one PR initially.
> 
> Yes. Your way is much better. But we suppose the issue of '-Wimplicit-int-float-conversion' is trivial and putting them in separate PRs might raise another internal review process (for our side) by which extra time is needed. I was wondering could we continue in one single PR. :)

Will split this PR.
In this PR, we focus on the warnings caused by -Wdeprecated-copy.
Will update the code soon. Will create a new PR to address JDK-8259288.

shqking commented:
> Thanks for your comments. @kimbarrett and @navyxliu
> I updated the patch based on my understanding. Please check the latest commit.
> 
> As @kimbarrett mentioned, I suppose there still exist the following problems to be addressed.
> 
> 1. why clang-10 with '-Wdeprecated-copy' does NOT raise warning for class DUIterator. It's weird.
> 2. the assert failure when building with gcc '-fno-elide-constructors'. Might not be related to our patch. (JDK-8259036)
> 3. the implementation of 'operator=' for class DUIterator_Fast might be problematic.

@kimbarrett 
Regarding problem 1, I believe this is because there exists use-defined dtor for class DUIterator and the deprecated copy ctor warning message ought to be emitted by '-Wdeprecated-copy-dtor' (which is not enabled by '-Wall' or '-Wextra').

Please refer to https://github.com/llvm/llvm-project/blob/release/10.x/clang/lib/Sema/SemaDeclCXX.cpp#L13585
The main logic is user-defined dtor/copy constructor/assignment operator is checked in order when diagnosing implicit-defined copy constructor, and warning message will be emitted (lines 13619 to 13625).

In this case of class DUIterator, user-defined dtor is provided but '-Wdeprecated-copy-dtor' is not enabled, clang-10 does not further check whether '-Wdeprecated-copy' is on or not.

I further checked 
1) I built openjdk with "--with-extra-cxxflags='-Wdeprecated-copy-dtor'", but the building failed earlier before class DUIterator.
2) I removed the dtor of class DUIterator manually and built openjdk with clang-10. Then clang-10 would produce the warning as we expected, which I think verified my conjecture.
```
=== Output from failing command(s) repeated here ===
* For target hotspot_variant-server_libjvm_gtest_objs_test_mathexact.o:
In file included from /home/haosun01/jdk/jdk_src/test/hotspot/gtest/opto/test_mathexact.cpp:26:
In file included from /home/haosun01/jdk/jdk_src/src/hotspot/share/opto/mulnode.hpp:28:
/home/haosun01/jdk/jdk_src/src/hotspot/share/opto/node.hpp:1342:8: error: definition of implicit copy constructor for 'DUIterator' is deprecated because it has a user-declared copy assignment operator [-Werror,-Wdeprecated-copy]
  void operator=(const DUIterator& that)
       ^
/home/haosun01/jdk/jdk_src/src/hotspot/share/opto/node.hpp:1347:12: note: in implicit copy constructor for 'DUIterator' first required here
  { return DUIterator(this, 0); }
           ^
/home/haosun01/jdk/jdk_src/src/hotspot/share/opto/node.hpp:1410:8: error: definition of implicit copy constructor for 'DUIterator_Fast' is deprecated because it has a user-declared copy assignment operator [-Werror,-Wdeprecated-copy]
  void operator=(const DUIterator_Fast& that)
       ^
/home/haosun01/jdk/jdk_src/src/hotspot/share/opto/node.hpp:1418:10: note: in implicit copy constructor for 'DUIterator_Fast' first required here
  return DUIterator_Fast(this, 0);
         ^
/home/haosun01/jdk/jdk_src/src/hotspot/share/opto/node.hpp:1467:8: error: definition of implicit copy constructor for 'DUIterator_Last' is deprecated because it has a user-declared copy assignment operator [-Werror,-Wdeprecated-copy]
   ... (rest of output omitted)
* For target support_gensrc_jdk.localedata__cldr-gensrc.marker:

* All command lines available in /home/haosun01/tmp/deprecated-copy/clang10-fast-build/make-support/failure-logs.
=== End of repeated output ===
```

shqking commented:
> Thanks for your comments. @kimbarrett and @navyxliu
> I updated the patch based on my understanding. Please check the latest commit.
> 
> As @kimbarrett mentioned, I suppose there still exist the following problems to be addressed.
> 
> 1. why clang-10 with '-Wdeprecated-copy' does NOT raise warning for class DUIterator. It's weird.
> 2. the assert failure when building with gcc '-fno-elide-constructors'. Might not be related to our patch. (JDK-8259036)
> 3. the implementation of 'operator=' for class DUIterator_Fast might be problematic.

For problem 1, it's due to '-Wdeprecated-copy-dtor'. Please refer to my previous comment for details.
For problem 2, I suppose it's not related to our patch and it would be issued in JDK-8259036.
For problem 3, as clarified by @vnkozlov , the code is correct.

Besides, the pre-submit tests failed due to one GC problem, which seems not related to this patch. See JDK-8258481.

Hence, I wonder if this patch is ready to merge. 
Could you please take another round of review as new commits have been made? @kimbarrett @navyxliu @vnkozlov 
Thanks.

vnkozlov reviewed (approved):
Still good.

kimbarrett commented on src/hotspot/share/opto/node.hpp:
```diff
@@ -1452,6 +1458,8 @@ class DUIterator_Last : private DUIterator_Fast {
   DUIterator_Last() { }
   // initialize to garbage
 
+  DUIterator_Last (const DUIterator_Last& that) = default;
```
Nit - remove space before parameter list.

shqking replied:
Thanks for point it out. Will remove.

kimbarrett reviewed (changes requested):
[Can't comment on this inline.] I'd prefer DUIterator_Last::operator= be changed to =default, for consistency with the copy constructor.  That would require fixing the return type too.

shqking commented:
> [Can't comment on this inline.] I'd prefer DUIterator_Last::operator= be changed to =default, for consistency with the copy constructor. That would require fixing the return type too.

Thanks for your comment. Agree. Will update the code.

shqking commented:
@vnkozlov would you mind helping to review the latest commit (again)? Really appreciate it.

vnkozlov reviewed (approved):
Looks good.

shqking commented:
Thanks a lot for your review @kimbarrett @vnkozlov .

/integrate

nsjian commented:
/sponsor
